src/Compilers/Core/Portable/Text/ChangedText.cs

```diff
@@ -83,8 +83,16 @@ public override void CopyTo(int sourceIndex, char[] destination, int destination
         public override SourceText WithChanges(IEnumerable<TextChange> changes)
         {
             // compute changes against newText to avoid capturing strong references to this ChangedText instance.
-            var changed = (ChangedText)_newText.WithChanges(changes);
-            return new ChangedText(this, changed._newText, changed._changes);
+            // note: _newText is never a ChangeText, so if WithChanges is a no-op the result will not be a ChangedText
+            var changed = _newText.WithChanges(changes) as ChangedText;  
+            if (changed != null)
+            {
+                return new ChangedText(this, changed._newText, changed._changes);
+            }
+            else 
+            {
+                return this;
```

Comment:
@CyrusNajmabadi Because I know what implementations of SourceText that _newText can have, and they will either return a ChangedText if they are changed, or they will return their own instance if there is no change.